lldb/source/Plugins/Process/Dpu/DpuRank.cpp

```diff
@@ -278,12 +406,47 @@ StateType Dpu::StepThread(uint32_t thread_index, unsigned int *exit_status) {
 
   int ret = DPU_API_SUCCESS;
   if (registers_has_been_modified) {
-    ret |= dpu_restore_context_for_dpu(m_dpu, m_context);
+    ret = dpu_restore_context_for_dpu(m_dpu, m_context);
+    if (ret != DPU_API_SUCCESS)
+      return StateType::eStateCrashed;
     registers_has_been_modified = false;
   }
-  ret |=
+
+  uint64_t inst_to_restore_if_printf_enable = UNKNOWN_INSTRUCTION;
+  uint32_t current_pc;
+  if (printf_enable) {
+    uint64_t inst_to_replace_with = UNKNOWN_INSTRUCTION;
+    current_pc = (m_context->pcs[thread_index] * 8);
+
+    if (!PrepareStepOverPrintfBkp(thread_index,
+                                  inst_to_restore_if_printf_enable,
+                                  inst_to_replace_with, current_pc))
+      return StateType::eStateCrashed;
+
+    if (inst_to_replace_with != UNKNOWN_INSTRUCTION) {
+      // Read what should be the bkp instruction and write the expected
+      // instruction instead
+      if (!ReadIRAM(current_pc, &inst_to_restore_if_printf_enable,
+                    sizeof(inst_to_restore_if_printf_enable)))
+        return StateType::eStateCrashed;
+      if (!WriteIRAM(current_pc, (const void *)&inst_to_replace_with,
```

Comment:
at some point, it might be interesting to overload WriteIRAM with a version taking directly a dpu_instruction_t or maybe better a vector of instructions